### ISSUE TITLE
MultibodyPlant should check Context validity

### ISSUE THREAD

**Issue by avalenzu** [RobotLocomotion/drake#11252]
Passing a diagram context into `Frame::CalcPose()` results in the following assertion:
```
abort: Failure at bazel-out/k8-dbg/bin/external/drake/systems/framework/_virtual_includes/cache_and_dependency_tracker/drake/systems/framework/cache.h:702 in get_cache_entry_value():
condition 'has_cache_entry_value(index)' failed.
```
I spent a fair few minutes wondering what had gone wrong in my diagram before I realized that the issue was a context mismatch. If the assertion had come from a `MultibodyPlant` method this would have been more apparent. We could add
```
DRAKE_ASSERT_VOID(System<double>::CheckValidContext(context));
```
in user-facing non-systems-framework methods of `MultibodyPlant`.

**Comment by amcastro-tri** [comment]
would you prefer ASSERT (only debug) or DEMAND?

**Comment by sherm1** [comment]
Context-validating is potentially very expensive.

**Comment by avalenzu** [comment]
I think ASSERT, definitely.

**Comment by amcastro-tri** [comment]
See related issue #11801.

**Comment by amcastro-tri** [comment]
Should this be an MBP feature or a `System` feature? @sherm1?

**Comment by jwnimmer-tri** [comment]
The check proposed by #12564 should never be disarmed (i.e., not only in Debug builds).  For the "correctly-mated System vs Context" checking, System methods (like EvalTimeDerivatives) will embed the check.  For MbP-specific functions like GetFreeBodyPose, the MbP must call ValidateContext, because the framework can't do it.

**Comment by sherm1** [comment]
Agreed. Eventually you would probably hit a framework method that would barf on the mismatch, but the delayed detection would be less helpful for users. (Would be better if the ValidateContext method would report the name of the failing API though.)

**Comment by jwnimmer-tri** [comment]
I think that #14389 resolved this.

**Comment by jwnimmer-tri** [comment]
While #14389 resolved the API directly on MbP, there are several other public classes on the MbT-related side that still lack calls to `ValidateContxt`, e.g. `Frame::CalcPoseInWorld` as noted in https://github.com/RobotLocomotion/drake/issues/12560#issuecomment-786366018.

**Comment by amcastro-tri** [comment]
How expensive are those ValidateContext()? can they safely be placed on release builds? 
Also, I imagine we want them in MBP's public APIs, not MBT?

**Comment by jwnimmer-tri** [comment]
Per [its API Doc](https://drake.mit.edu/doxygen_cxx/classdrake_1_1systems_1_1_system_base.html#a7622a0fb573704de7505804d57e66a79), calling `ValidateContext` is sufficiently fast for performance sensitive code.

PR #14389 already added the checks to all of the MbP methods.

What's still needed here is to add it to any of othe `multibody/tree/...` methods that take a `Context`, e.g., `Frame::CalcPoseInWorld` and others.

**Comment by rpoyner-tri** [comment]
@amcastro-tri do you have any work-in-progress on this? If not, I think i have a patch that should do the job.

**Comment by amcastro-tri** [comment]
Your patch is excellent. Thanks @rpoyner-tri!

**Comment by jwnimmer-tri** [comment]
In #15489, I found that `MultibodyPlant::SetDefaultState` and `MultibodyPlant::SetRandomState` lack validation checks on the `State*` in/out pointer.  They call `CheckValidState`, but that function only does a basic "continuous xor discrete" sanity check, it doesn't inspect the SystemId tag.
